File: web_next/lib/subjects.ts
/**
 * 科目の固定リスト（順番も固定）
 * FastAPIのconfig/constants.pyのFIXED_SUBJECTSと一致させる
 */
export const FIXED_SUBJECTS = [
  "憲法",
  "行政法",
  "民法",
  "商法",
  "民事訴訟法",
  "刑法",
  "刑事訴訟法",
  "一般教養科目",
  "実務基礎（民事）",
  "実務基礎（刑事）",
  "労働法",
  "倒産法",
  "租税法",
  "経済法",
  "知的財産法",
  "環境法",
  "国際関係法（公法系）",
  "国際関係法（私法系）",
] as const

/**
 * 科目ID → 科目名のマッピング（1-18）
 * バックエンドのconfig/subjects.pyのSUBJECT_MAPと一致させる
 */
export const SUBJECT_MAP: Record<number, string> = {
  1: "憲法",
  2: "行政法",
  3: "民法",
  4: "商法",
  5: "民事訴訟法",
  6: "刑法",
  7: "刑事訴訟法",
  8: "一般教養科目",
  9: "実務基礎（民事）",
  10: "実務基礎（刑事）",
  11: "労働法",
  12: "倒産法",
  13: "租税法",
  14: "経済法",
  15: "知的財産法",
  16: "環境法",
  17: "国際関係法（公法系）",
  18: "国際関係法（私法系）",
} as const

/**
 * 科目名 → 科目IDの逆マッピング
 */
export const SUBJECT_NAME_TO_ID: Record<string, number> = Object.fromEntries(
  Object.entries(SUBJECT_MAP).map(([id, name]) => [name, parseInt(id)])
) as Record<string, number>

/**
 * 科目IDから科目名を取得
 */
export function getSubjectName(subjectId: number | null | undefined): string {
  if (subjectId === null || subjectId === undefined) return "不明"
  return SUBJECT_MAP[subjectId] || "不明"
}

/**
 * 科目名から科目IDを取得
 */
export function getSubjectId(subjectName: string): number | null {
  return SUBJECT_NAME_TO_ID[subjectName] || null
}

/**
 * 科目IDが有効かチェック（1-18）
 */
export function isValidSubjectId(subjectId: number): boolean {
  return 1 <= subjectId && subjectId <= 18
}

/**
 * 科目リストを固定順序で並べ替える
 * @param subjects APIから取得した科目リスト
 * @returns 固定順序で並べ替えられた科目リスト
 */
export function sortSubjectsByFixedOrder(subjects: string[]): string[] {
  // FIXED_SUBJECTSの順序に従って並べ替え
  const sorted: string[] = []
  const subjectSet = new Set(subjects)
  
  // FIXED_SUBJECTSに含まれる科目を順序通りに追加
  for (const subject of FIXED_SUBJECTS) {
    if (subjectSet.has(subject)) {
      sorted.push(subject)
      subjectSet.delete(subject)
    }
  }
  
  // FIXED_SUBJECTSにない科目も追加（後方互換性）
  const additional = Array.from(subjectSet).sort()
  sorted.push(...additional)
  
  return sorted
}

File: config/subjects.py
"""
科目マッピング定義（1-18）
バックエンドとフロントエンドで共通の定義
"""

# 科目ID → 科目名のマッピング（1-18）
SUBJECT_MAP = {
    1: "憲法",
    2: "行政法",
    3: "民法",
    4: "商法",
    5: "民事訴訟法",
    6: "刑法",
    7: "刑事訴訟法",
    8: "一般教養科目",
    9: "実務基礎（民事）",
    10: "実務基礎（刑事）",
    11: "労働法",
    12: "倒産法",
    13: "租税法",
    14: "経済法",
    15: "知的財産法",
    16: "環境法",
    17: "国際関係法（公法系）",
    18: "国際関係法（私法系）",
}

# 科目名 → 科目IDの逆マッピング
SUBJECT_NAME_TO_ID = {name: id for id, name in SUBJECT_MAP.items()}

# 科目IDのリスト（表示順序）
SUBJECT_IDS = list(SUBJECT_MAP.keys())

# 科目名のリスト（表示順序、FIXED_SUBJECTSと同じ）
SUBJECT_NAMES = list(SUBJECT_MAP.values())


def get_subject_name(subject_id: int) -> str:
    """科目IDから科目名を取得"""
    return SUBJECT_MAP.get(subject_id, "不明")


def get_subject_id(subject_name: str) -> int | None:
    """科目名から科目IDを取得"""
    if not subject_name:
        return None
    # "商 法" のような空白混入や "(民事)" のような半角括弧を吸収
    s = "".join(str(subject_name).split())
    s = s.replace("(", "（").replace(")", "）")
    v = SUBJECT_NAME_TO_ID.get(s)
    if v is not None:
        return v

    # かっこ無しの揺れ（例: "実務基礎民事"）
    s2 = s.replace("（", "").replace("）", "")
    for name, sid in SUBJECT_NAME_TO_ID.items():
        if name.replace("（", "").replace("）", "") == s2:
            return sid
    return None


def is_valid_subject_id(subject_id: int) -> bool:
    """科目IDが有効かチェック（1-18）"""
    return 1 <= subject_id <= 18
